cmd/update-main.go

```diff
@@ -17,274 +17,478 @@
+		}
+		return fmt.Sprintf
+	}()
+)
+
+func doUpdate(sha256Hex string, latestReleaseTime time.Time, ok bool) (updateStatusMsg string, err *probe.Error) {
+	if !ok {
+		updateStatusMsg = colorGreenBold("mc update to version RELEASE.%s canceled.",
+			latestReleaseTime.Format(mcReleaseTagTimeLayout))
+		return updateStatusMsg, nil
+	}
+	var sha256Sum []byte
+	var e error
+	sha256Sum, e = hex.DecodeString(sha256Hex)
+	if e != nil {
+		return updateStatusMsg, probe.NewError(e)
+	}
+
+	resp, e := http.Get(getDownloadURL(releaseTimeToReleaseTag(latestReleaseTime)))
+	if e != nil {
+		return updateStatusMsg, probe.NewError(e)
 	}
+	defer resp.Body.Close()
 
-	return older, downloadURL, nil
+	// FIXME: add support for gpg verification as well.
+	if e = update.Apply(resp.Body,
+		update.Options{
+			Hash:     crypto.SHA256,
+			Checksum: sha256Sum,
+		},
+	); e != nil {
+		return updateStatusMsg, probe.NewError(e)
+	}
+
+	return colorGreenBold("mc updated to version RELEASE.%s successfully.",
+		latestReleaseTime.Format(mcReleaseTagTimeLayout)), nil
+}
+
+func shouldUpdate(quiet bool, sha256Hex string, latestReleaseTime time.Time) (ok bool) {
+	ok = true
+	if !quiet {
+		ok = prompt.Confirm(colorGreenBold("Update to RELEASE.%s [%s]", latestReleaseTime.Format(mcReleaseTagTimeLayout), "yes"))
+	}
+	return ok
+}
+
+type updateMessage struct {
+	Status  string `json:"status"`
+	Message string `json:"message"`
+}
+
+// String colorized make bucket message.
+func (s updateMessage) String() string {
+	return s.Message
+}
+
+// JSON jsonified make bucket message.
+func (s updateMessage) JSON() string {
+	updateJSONBytes, e := json.Marshal(s)
+	fatalIf(probe.NewError(e), "Unable to marshal into JSON.")
+
+	return string(updateJSONBytes)
 }
 
 func mainUpdate(ctx *cli.Context) {
 	if len(ctx.Args()) != 0 {
 		cli.ShowCommandHelpAndExit(ctx, "update", -1)
 	}
 
-	// Additional command speific theme customization.
-	console.SetColor("Update", color.New(color.FgGreen, color.Bold))
+	quiet := ctx.Bool("quiet") || ctx.GlobalBool("quiet")
 
-	older, downloadURL, err := getUpdateInfo(10 * time.Second)
-	fatalIf(err.Trace(downloadURL), "Unable to fetch update info for mc.")
+	updateMsg, sha256Hex, _, latestReleaseTime, err := getUpdateInfo(10 * time.Second)
+	if err != nil {
+		errorIf(err, "Unable to update ‘mc’.")
+		os.Exit(-1)
+	}
 
-	if !globalQuiet {
+	// Nothing to update running the latest release.
+	color.New(color.FgGreen, color.Bold)
+	if updateMsg == "" {
 		printMsg(updateMessage{
-			olderThan: older,
-			Download:  downloadURL,
-			Version:   Version,
+			Status:  "success",
+			Message: colorGreenBold("You are already running the most recent version of ‘mc’."),
 		})
+		os.Exit(0)
+	}
+
+	printMsg(updateMessage{
+		Status:  "success",
+		Message: updateMsg,
+	})
+
+	// if the in-place update is disabled then we shouldn't ask the
+	// user to update the binaries.
+	if strings.Contains(updateMsg, mcReleaseURL) {
```

Comment:
Not sure when in-place update is disabled, and what's the relation with this strings.Contains(), can you elaborate ?

Reply:
There is no reason to perform an update if its not a released binary.